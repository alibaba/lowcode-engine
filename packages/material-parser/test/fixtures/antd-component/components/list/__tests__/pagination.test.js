import React from 'react';
import { render, mount } from 'enzyme';
import List from '..';

describe('List.pagination', () => {
  const data = [
    { key: 0, name: 'Jack' },
    { key: 1, name: 'Lucy' },
    { key: 2, name: 'Tom' },
    { key: 3, name: 'Jerry' },
  ];

  const pagination = { className: 'my-page', pageSize: 2 };

  function createList(props) {
    return (
      <List
        itemLayout="vertical"
        pagination={pagination}
        dataSource={data}
        renderItem={item => <List.Item key={item.key}>{item.name}</List.Item>}
        {...props}
      />
    );
  }

  function renderedNames(wrapper) {
    return wrapper.find('.ant-list-item').map(row => row.text());
  }

  it('renders pagination correctly', () => {
    const wrapper = render(createList());
    expect(wrapper).toMatchSnapshot();
  });

  it('should not show pager if pagination.hideOnSinglePage is true and only 1 page', () => {
    const wrapper = mount(createList({ pagination: { pageSize: 3, hideOnSinglePage: true } }));
    expect(wrapper.find('.ant-pagination')).toHaveLength(1);
    wrapper.setProps({ pagination: { pageSize: 3, hideOnSinglePage: false } });
    expect(wrapper.find('.ant-pagination')).toHaveLength(1);
    wrapper.setProps({ pagination: { pageSize: 4, hideOnSinglePage: true } });
    expect(wrapper.find('.ant-pagination')).toHaveLength(0);
    wrapper.setProps({ pagination: { pageSize: 4, hideOnSinglePage: false } });
    expect(wrapper.find('.ant-pagination')).toHaveLength(1);
    wrapper.setProps({ pagination: { pageSize: 5, hideOnSinglePage: true } });
    expect(wrapper.find('.ant-pagination')).toHaveLength(0);
    wrapper.setProps({ pagination: { pageSize: 5, hideOnSinglePage: false } });
    expect(wrapper.find('.ant-pagination')).toHaveLength(1);
  });

  it('paginate data', () => {
    const wrapper = mount(createList());

    expect(renderedNames(wrapper)).toEqual(['Jack', 'Lucy']);
    wrapper
      .find('Pager')
      .last()
      .simulate('click');
    expect(renderedNames(wrapper)).toEqual(['Tom', 'Jerry']);
  });

  it('repaginates when pageSize change', () => {
    const wrapper = mount(createList());

    wrapper.setProps({ pagination: { pageSize: 1 } });
    expect(renderedNames(wrapper)).toEqual(['Jack']);
  });

  it('fires change event', () => {
    const handlePaginationChange = jest.fn();
    const noop = () => {};
    const wrapper = mount(
      createList({
        pagination: {
          ...pagination,
          onChange: handlePaginationChange,
          onShowSizeChange: noop,
        },
      }),
    );

    wrapper
      .find('Pager')
      .last()
      .simulate('click');

    expect(handlePaginationChange).toHaveBeenCalledWith(2, 2);
  });

  // https://github.com/ant-design/ant-design/issues/4532
  // https://codepen.io/afc163/pen/pWVRJV?editors=001
  it('should display pagination as prop pagination change between true and false', () => {
    const wrapper = mount(createList());
    expect(wrapper.find('.ant-pagination')).toHaveLength(1);
    expect(wrapper.find('.ant-pagination-item')).toHaveLength(2);
    wrapper.setProps({ pagination: false });
    expect(wrapper.find('.ant-pagination')).toHaveLength(0);
    wrapper.setProps({ pagination });
    wrapper.update();
    expect(wrapper.find('.ant-pagination')).toHaveLength(1);
    expect(wrapper.find('.ant-pagination-item')).toHaveLength(2);
    wrapper.find('.ant-pagination-item-2').simulate('click');
    expect(renderedNames(wrapper)).toEqual(['Tom', 'Jerry']);
    wrapper.setProps({ pagination: false });
    expect(wrapper.find('.ant-pagination')).toHaveLength(0);
    wrapper.setProps({ pagination: true });
    expect(wrapper.find('.ant-pagination')).toHaveLength(1);
    // Legacy code will make pageSize ping with 10, here we fixed to keep sync by current one
    expect(wrapper.find('.ant-pagination-item')).toHaveLength(2);
    expect(renderedNames(wrapper)).toEqual(['Tom', 'Jerry']);
  });

  // https://github.com/ant-design/ant-design/issues/5259
  it('change to correct page when data source changes', () => {
    const wrapper = mount(createList({ pagination: { pageSize: 1 } }));
    wrapper.find('.ant-pagination-item-3').simulate('click');
    wrapper.setProps({ dataSource: [data[0]] });
    expect(wrapper.find('.ant-pagination-item-1').hasClass('ant-pagination-item-active')).toBe(
      true,
    );
  });

  it('specify the position of pagination', () => {
    const wrapper = mount(createList({ pagination: { position: 'top' } }));
    expect(
      wrapper
        .find('.ant-list')
        .childAt(0)
        .find('.ant-pagination'),
    ).toHaveLength(1);
    wrapper.setProps({ pagination: { position: 'bottom' } });
    expect(
      wrapper
        .find('.ant-list')
        .children()
        .last()
        .find('.ant-pagination'),
    ).toHaveLength(1);
    wrapper.setProps({ pagination: { position: 'both' } });
    expect(wrapper.find('.ant-pagination')).toHaveLength(2);
    expect(
      wrapper
        .find('.ant-list')
        .childAt(0)
        .find('.ant-pagination'),
    ).toHaveLength(1);
    expect(
      wrapper
        .find('.ant-list')
        .children()
        .last()
        .find('.ant-pagination'),
    ).toHaveLength(1);
  });

  it('should change page size work', () => {
    const wrapper = mount(createList({ pagination: { showSizeChanger: true } }));
    expect(
      wrapper
        .find('Pagination')
        .first()
        .render(),
    ).toMatchSnapshot();

    wrapper.find('.ant-select-selector').simulate('mousedown');
    wrapper
      .find('.ant-select-item-option')
      .at(2)
      .simulate('click');

    wrapper.find('.ant-select-selector').simulate('mousedown');
    expect(
      wrapper
        .find('Pagination')
        .first()
        .render(),
    ).toMatchSnapshot();
  });

  it('should default work', () => {
    const wrapper = mount(
      createList({
        pagination: {
          defaultPageSize: 3,
          defaultCurrent: 2,
          pageSizeOptions: ['1', '3'],
          showSizeChanger: true,
        },
      }),
    );

    expect(
      wrapper
        .find('Pagination')
        .first()
        .render(),
    ).toMatchSnapshot();
  });

  it('should not crash when pagination is null', () => {
    mount(
      createList({
        pagination: null,
      }),
    );
  });
});
