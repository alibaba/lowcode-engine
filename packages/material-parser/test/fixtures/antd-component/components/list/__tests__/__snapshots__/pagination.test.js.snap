// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List.pagination renders pagination correctly 1`] = `
<div
  class="ant-list ant-list-vertical ant-list-split ant-list-something-after-last-item"
>
  <div
    class="ant-spin-nested-loading"
  >
    <div
      class="ant-spin-container"
    >
      <ul
        class="ant-list-items"
      >
        <li
          class="ant-list-item ant-list-item-no-flex"
        >
          Jack
        </li>
        <li
          class="ant-list-item ant-list-item-no-flex"
        >
          Lucy
        </li>
      </ul>
    </div>
  </div>
  <div
    class="ant-list-pagination"
  >
    <ul
      class="ant-pagination my-page"
      unselectable="unselectable"
    >
      <li
        aria-disabled="true"
        class="ant-pagination-prev ant-pagination-disabled"
        title="Previous Page"
      >
        <a
          class="ant-pagination-item-link"
          disabled=""
        >
          <span
            aria-label="left"
            class="anticon anticon-left"
            role="img"
          >
            <svg
              aria-hidden="true"
              class=""
              data-icon="left"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
              />
            </svg>
          </span>
        </a>
      </li>
      <li
        class="ant-pagination-item ant-pagination-item-1 ant-pagination-item-active"
        tabindex="0"
        title="1"
      >
        <a>
          1
        </a>
      </li>
      <li
        class="ant-pagination-item ant-pagination-item-2"
        tabindex="0"
        title="2"
      >
        <a>
          2
        </a>
      </li>
      <li
        aria-disabled="false"
        class="ant-pagination-next"
        tabindex="0"
        title="Next Page"
      >
        <a
          class="ant-pagination-item-link"
        >
          <span
            aria-label="right"
            class="anticon anticon-right"
            role="img"
          >
            <svg
              aria-hidden="true"
              class=""
              data-icon="right"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
              />
            </svg>
          </span>
        </a>
      </li>
    </ul>
  </div>
</div>
`;

exports[`List.pagination should change page size work 1`] = `
<ul
  class="ant-pagination"
  unselectable="unselectable"
>
  <li
    aria-disabled="true"
    class="ant-pagination-prev ant-pagination-disabled"
    title="Previous Page"
  >
    <a
      class="ant-pagination-item-link"
      disabled=""
    >
      <span
        aria-label="left"
        class="anticon anticon-left"
        role="img"
      >
        <svg
          aria-hidden="true"
          class=""
          data-icon="left"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
          />
        </svg>
      </span>
    </a>
  </li>
  <li
    class="ant-pagination-item ant-pagination-item-1 ant-pagination-item-active"
    tabindex="0"
    title="1"
  >
    <a>
      1
    </a>
  </li>
  <li
    aria-disabled="true"
    class="ant-pagination-next ant-pagination-disabled"
    title="Next Page"
  >
    <a
      class="ant-pagination-item-link"
      disabled=""
    >
      <span
        aria-label="right"
        class="anticon anticon-right"
        role="img"
      >
        <svg
          aria-hidden="true"
          class=""
          data-icon="right"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
          />
        </svg>
      </span>
    </a>
  </li>
  <li
    class="ant-pagination-options"
  >
    <div
      class="ant-select ant-pagination-options-size-changer ant-select-single ant-select-show-arrow"
    >
      <div
        class="ant-select-selector"
      >
        <span
          class="ant-select-selection-search"
        >
          <input
            aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
            aria-autocomplete="list"
            aria-controls="rc_select_TEST_OR_SSR_list"
            aria-haspopup="listbox"
            aria-owns="rc_select_TEST_OR_SSR_list"
            autocomplete="off"
            class="ant-select-selection-search-input"
            id="rc_select_TEST_OR_SSR"
            readonly=""
            role="combobox"
            style="opacity: 0;"
            value=""
          />
        </span>
        <span
          class="ant-select-selection-item"
        >
          10 / page
        </span>
      </div>
      <span
        aria-hidden="true"
        class="ant-select-arrow"
        style="user-select: none;"
        unselectable="on"
      >
        <span
          aria-label="down"
          class="anticon anticon-down"
          role="img"
        >
          <svg
            aria-hidden="true"
            class=""
            data-icon="down"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
            />
          </svg>
        </span>
      </span>
    </div>
  </li>
</ul>
`;

exports[`List.pagination should change page size work 2`] = `
<ul
  class="ant-pagination"
  unselectable="unselectable"
>
  <li
    aria-disabled="true"
    class="ant-pagination-prev ant-pagination-disabled"
    title="Previous Page"
  >
    <a
      class="ant-pagination-item-link"
      disabled=""
    >
      <span
        aria-label="left"
        class="anticon anticon-left"
        role="img"
      >
        <svg
          aria-hidden="true"
          class=""
          data-icon="left"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
          />
        </svg>
      </span>
    </a>
  </li>
  <li
    class="ant-pagination-item ant-pagination-item-1 ant-pagination-item-active"
    tabindex="0"
    title="1"
  >
    <a>
      1
    </a>
  </li>
  <li
    aria-disabled="true"
    class="ant-pagination-next ant-pagination-disabled"
    title="Next Page"
  >
    <a
      class="ant-pagination-item-link"
      disabled=""
    >
      <span
        aria-label="right"
        class="anticon anticon-right"
        role="img"
      >
        <svg
          aria-hidden="true"
          class=""
          data-icon="right"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
          />
        </svg>
      </span>
    </a>
  </li>
  <li
    class="ant-pagination-options"
  >
    <div
      class="ant-select ant-pagination-options-size-changer ant-select-single ant-select-show-arrow ant-select-open"
    >
      <div
        class="ant-select-selector"
      >
        <span
          class="ant-select-selection-search"
        >
          <input
            aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
            aria-autocomplete="list"
            aria-controls="rc_select_TEST_OR_SSR_list"
            aria-expanded="true"
            aria-haspopup="listbox"
            aria-owns="rc_select_TEST_OR_SSR_list"
            autocomplete="off"
            class="ant-select-selection-search-input"
            id="rc_select_TEST_OR_SSR"
            readonly=""
            role="combobox"
            style="opacity: 0;"
            value=""
          />
        </span>
        <span
          class="ant-select-selection-item"
        >
          50 / page
        </span>
      </div>
      <div>
        <div
          class="ant-select-dropdown"
          style="min-width: 0; opacity: 0;"
        >
          <div>
            <div
              id="rc_select_TEST_OR_SSR_list"
              role="listbox"
              style="height: 0px; overflow: hidden;"
            >
              <div
                aria-selected="false"
                id="rc_select_TEST_OR_SSR_list_0"
                role="option"
              >
                10
              </div>
              <div
                aria-selected="false"
                id="rc_select_TEST_OR_SSR_list_1"
                role="option"
              >
                20
              </div>
            </div>
            <div
              class=""
              style="height: 256px;"
            >
              <div>
                <div
                  class=""
                  style="display: flex; flex-direction: column;"
                >
                  <div
                    aria-selected="false"
                    class="ant-select-item ant-select-item-option ant-select-item-option-active"
                  >
                    <div
                      class="ant-select-item-option-content"
                    >
                      10 / page
                    </div>
                    <span
                      aria-hidden="true"
                      class="ant-select-item-option-state"
                      style="user-select: none;"
                      unselectable="on"
                    />
                  </div>
                  <div
                    aria-selected="false"
                    class="ant-select-item ant-select-item-option"
                  >
                    <div
                      class="ant-select-item-option-content"
                    >
                      20 / page
                    </div>
                    <span
                      aria-hidden="true"
                      class="ant-select-item-option-state"
                      style="user-select: none;"
                      unselectable="on"
                    />
                  </div>
                  <div
                    aria-selected="true"
                    class="ant-select-item ant-select-item-option ant-select-item-option-selected"
                  >
                    <div
                      class="ant-select-item-option-content"
                    >
                      50 / page
                    </div>
                    <span
                      aria-hidden="true"
                      class="ant-select-item-option-state"
                      style="user-select: none;"
                      unselectable="on"
                    />
                  </div>
                  <div
                    aria-selected="false"
                    class="ant-select-item ant-select-item-option"
                  >
                    <div
                      class="ant-select-item-option-content"
                    >
                      100 / page
                    </div>
                    <span
                      aria-hidden="true"
                      class="ant-select-item-option-state"
                      style="user-select: none;"
                      unselectable="on"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <span
        aria-hidden="true"
        class="ant-select-arrow"
        style="user-select: none;"
        unselectable="on"
      >
        <span
          aria-label="down"
          class="anticon anticon-down"
          role="img"
        >
          <svg
            aria-hidden="true"
            class=""
            data-icon="down"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
            />
          </svg>
        </span>
      </span>
      <div
        style="position: absolute; top: 0px; left: 0px; width: 100%;"
      />
      <div
        style="position: absolute; top: 0px; left: 0px; width: 100%;"
      />
      <div
        style="position: absolute; top: 0px; left: 0px; width: 100%;"
      />
      <div
        style="position: absolute; top: 0px; left: 0px; width: 100%;"
      />
      <div
        style="position: absolute; top: 0px; left: 0px; width: 100%;"
      />
    </div>
  </li>
</ul>
`;

exports[`List.pagination should default work 1`] = `
<ul
  class="ant-pagination"
  unselectable="unselectable"
>
  <li
    aria-disabled="false"
    class="ant-pagination-prev"
    tabindex="0"
    title="Previous Page"
  >
    <a
      class="ant-pagination-item-link"
    >
      <span
        aria-label="left"
        class="anticon anticon-left"
        role="img"
      >
        <svg
          aria-hidden="true"
          class=""
          data-icon="left"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
          />
        </svg>
      </span>
    </a>
  </li>
  <li
    class="ant-pagination-item ant-pagination-item-1"
    tabindex="0"
    title="1"
  >
    <a>
      1
    </a>
  </li>
  <li
    class="ant-pagination-item ant-pagination-item-2 ant-pagination-item-active"
    tabindex="0"
    title="2"
  >
    <a>
      2
    </a>
  </li>
  <li
    aria-disabled="true"
    class="ant-pagination-next ant-pagination-disabled"
    title="Next Page"
  >
    <a
      class="ant-pagination-item-link"
      disabled=""
    >
      <span
        aria-label="right"
        class="anticon anticon-right"
        role="img"
      >
        <svg
          aria-hidden="true"
          class=""
          data-icon="right"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
          />
        </svg>
      </span>
    </a>
  </li>
  <li
    class="ant-pagination-options"
  >
    <div
      class="ant-select ant-pagination-options-size-changer ant-select-single ant-select-show-arrow"
    >
      <div
        class="ant-select-selector"
      >
        <span
          class="ant-select-selection-search"
        >
          <input
            aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
            aria-autocomplete="list"
            aria-controls="rc_select_TEST_OR_SSR_list"
            aria-haspopup="listbox"
            aria-owns="rc_select_TEST_OR_SSR_list"
            autocomplete="off"
            class="ant-select-selection-search-input"
            id="rc_select_TEST_OR_SSR"
            readonly=""
            role="combobox"
            style="opacity: 0;"
            value=""
          />
        </span>
        <span
          class="ant-select-selection-item"
        >
          3 / page
        </span>
      </div>
      <span
        aria-hidden="true"
        class="ant-select-arrow"
        style="user-select: none;"
        unselectable="on"
      >
        <span
          aria-label="down"
          class="anticon anticon-down"
          role="img"
        >
          <svg
            aria-hidden="true"
            class=""
            data-icon="down"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
            />
          </svg>
        </span>
      </span>
    </div>
  </li>
</ul>
`;
